src/groupeddataframe/utils.jl

```diff
@@ -323,54 +341,67 @@ function row_group_slots(cols::NTuple{N, AbstractVector},
             end
             refmap
         end
-        @inbounds @spawn_for_chunks 1_000_000 for i in eachindex(groups)
-            local refs_i
-            let i=i # Workaround for julia#15276
-                refs_i = map(c -> c[i], refarrays)
-            end
-            vals = map((m, r, s, fi) -> m[r-fi+1] * s, refmaps, refs_i, strides, firstinds)
-            j = sum(vals) + 1
-            # x < 0 happens with -1 in refmap, which corresponds to missing
-            if skipmissing && any(x -> x < 0, vals)
-                j = 0
-            else
-                seen[j] = true
+        @spawn_for_chunks chunk_size for i in eachindex(groups)
+            @inbounds begin
+                seeni = seen[Threads.threadid()]
```

Comment:
I think the compiler optimizes this out (actually this is something you were convincing me some time ago that happens in such situations 😄, but I will investigate).

Reply:
I know it generally happens with simple `if` inside loops, but I'm not sure that also applies to array indexing. I would expect it to be hard to reason about for the compiler, as `seen` may be mutated e.g. by another thread.

Reply:
I will check

Reply:
When I do what you propose then Julia starts boxing `seeni` and I could not force it not to do boxing. However, the comparison of performance against 1.0.1 release seems to show that the compiler is able to handle it correctly.

Reply:
Actually it seems it is better to have this operation inside the loop. I do not understand why as it does not make much sense (I do not understand something most likely):
```
julia> function f()
           x = [[0.0] for i in 1:Threads.nthreads()]
           @sync for i in 1:100
               Threads.@spawn for j in 1:100_000
                   @inbounds y = x[Threads.threadid()]
                   @inbounds y[1] += sin(j)
               end
           end
           return x
       end
f (generic function with 1 method)

julia> function g()
           x = [[0.0] for i in 1:Threads.nthreads()]
           @sync for i in 1:100
               Threads.@spawn begin
                   @inbounds y = x[Threads.threadid()]
                   for j in 1:100_000
                       @inbounds y[1] += sin(j)
                   end
               end
           end
           return x
       end
g (generic function with 1 method)

julia> @btime f()
  65.672 ms (622 allocations: 48.38 KiB)
4-element Vector{Vector{Float64}}:
 [72.06330704159137]
 [25.868879450822345]
 [53.585536005280694]
 [33.25998786534185]

julia> @btime g()
  70.029 ms (622 allocations: 48.38 KiB)
4-element Vector{Vector{Float64}}:
 [53.585536005280694]
 [53.585536005280694]
 [25.868879450822345]
 [51.73775890164989]
```

@tkf: any idea?

Reply:
Maybe try `@spawn let seen=seen` to prevent boxing?

Reply:
Note that combining `@spawn` and `scratches[Threads.threadid()]` pattern is not correct (almost always) or at least very dangerous.  This is because it'd be a data race if your code yield to the scheduler (and there is no way to suppress this ATM) while mutating `seen[Threads.threadid()]`.  This can easily happen during, e.g., print-based debugging

```julia
seeni = seen[Threads.threadid()]
...

@show some_variable  # for debugging (maybe in some function called from here)

...
use(seeni)
```

Here is a demo

```julia
julia> function f(x)
           @show x  # it works if you comment out this line
           return x
       end

       function demo(repeat = 10)
           xs = zeros(Int, Threads.nthreads())
           @sync for i in 1:repeat
               Threads.@spawn begin
                   xs[Threads.threadid()] += f(i)
               end
           end
           xs
       end

       xs = demo()
x = 3
x = 7
x = 5
x = 8
x = 1
x = 4
x = 2
x = 6
x = 9
x = 10
4-element Vector{Int64}:
 10
  7
  8
  2

julia> sum(xs)
27

julia> sum(1:10)
55
```

I discussed it a bit also in https://juliafolds.github.io/FLoops.jl/dev/explanation/faq/#faq-state-threadid.

You can avoid it by making sure different task cannot access to different `seen` arrays, by construction:

```julia
seen_buffers = [BitVector(...) for _ in 1:nthreads()]
@assert length(other_inputs) == nthreads()  # suppose already chunked
Threads.@sync for (seen, input) in zip(seen_buffers, other_inputs)
    Threads.@spawn begin
        for i in inputs
            process!(seen, i)
        end
    end
end
seen = seen_buffers[1]
for other in seen_buffers[2:end]
    seen .|= other
end
```

or use divide-and-conquer approach, to parallelize the merges of `seen_buffers`, as discussed in slack

Reply:
OK. Switching to `Threads.@threads` then. Thank you!